### ISSUE TITLE
Consider binding llama.cpp directly?

### ISSUE THREAD

**Issue by shadowmint** [rustformers/llama-rs#124]
I love this project, but I wonder if it could be served by having a parallel binding for llama.cpp.

The justification for this is quite simple: this project is chasing a moving target.

Until llama.cpp settles down (the change rate is wild right now), how can we hope to keep in sync?

If we're always spending all the time trying to catch up with llama.cpp, then the rust version will always be a second best library, that people use reluctantly. 

Ultimately, and ideally, it would be a full implementation in rust with no C/C++ dependencies, but that seems like a distant dream at the moment...

### A way forward...

Since the cmake build profile was added to llama.cpp, it is now possible to build llama.cpp *directly as a dependency* and access the API in https://github.com/ggerganov/llama.cpp/blob/master/llama.h using bindgen.

Rather than re-implementing every change in llama.cpp this gives you *instant feature parity* with that library, and would allow people to get on with building an idiomatic rust api for it, instead of spending all the time and effort chasing the endless changes in llama.cpp. You effectively get 'for free' a whole lot of stuff, like gpt4all as it become available upstream.

Then this crate would become a **superset** of the functionality in llama.cpp, and that would be pretty cool, and let people focus of building cool stuff in rust instead of C++. 🦀 🦀 🦀

### Proof of concept

I know, asking for things is easy, doing this is hard, but this is actually *really easy* now. I've created a tiny PoC that demonstrates how trivial it is to setup here (hint: just set `BUILD_SHARED_LIBS` in cmake):
https://github.com/shadowmint/llama-sys/blob/main/build.rs

And and example of use here:
https://github.com/shadowmint/llama-sys/blob/main/tests/test_generate.rs

It runs perfectly:

```
running 1 test
llama_model_load: loading model from 'models/13B/model.bin' - please wait ...
llama_model_load: n_vocab = 32000
llama_model_load: n_ctx   = 512
llama_model_load: n_embd  = 5120
llama_model_load: n_mult  = 256
llama_model_load: n_head  = 40
llama_model_load: n_layer = 40
llama_model_load: n_rot   = 128
llama_model_load: f16     = 2
llama_model_load: n_ff    = 13824
llama_model_load: n_parts = 2
llama_model_load: type    = 2
llama_model_load: ggml map size = 7759.83 MB
llama_model_load: ggml ctx size = 101.25 KB
llama_model_load: mem required  = 9807.93 MB (+ 1608.00 MB per state)
llama_model_load: loading tensors from 'models/13B/model.bin'
llama_model_load: model size =  7759.39 MB / num tensors = 363
llama_init_from_file: kv self size  =  400.00 MB
loaded model
prompt length: 20
model context length: 512
prompt:
 hello, my name is dave and I live in Oz. Where do I live?

generating...

! I don't know, but it feels really good!!
I am an Aussie, not from the bush, although you might think so after hearing me talk, but I am a proud Aussie none-the-less!!
When my family and I decided to immigrate to Australia, we all knew that it was going to be hard at times; however, if you are willing to put in the effort then you can make it anywhere, even in this great land of Oz.
We left our old home behind, selling everything that wouldn't fit into a suitcase and moving as far
```

While I don't think this should replace the work on this repo (I do believe ultimately a full rust implementation is the best eventual goal), given that this project already depends on heavily on ggml, does it add much more overhead to have a binding to llama.cpp as well?

I think it's a tangible way forward that avoids the endless chasing of features and changes in llama.cpp.

Of course there are downsides... you become dependent on llama.cpp and subject to the changes and perf regressions in it, instead of an independent implementation that can go it's own way; so, if this is of no interest, feel free to just close this issue.

Thanks again for all your work on this crate, it's great!

**Comment by iacore** [comment]
uh... https://crates.io/crates/llama-sys

**Comment by KerfuffleV2** [comment]
> Until llama.cpp settles down (the change rate is wild right now), how can we hope to keep in sync?

> If we're always spending all the time trying to catch up with llama.cpp, then the rust version will always be a second best library, that people use reluctantly.

Is it really as bad as all that? It seems like most of the meaningful changes are happening in GGML rather than `llama.cpp` itself. The thing that really hurts is all the random undocumented file format changes.

**Comment by shadowmint** [comment]
The llama.cpp roadmap (https://github.com/ggerganov/llama.cpp/discussions/784) seems like it has more than just model format changes in it to me.

...but, if you don't think it's worth it, by all means, feel free to close this.

**Comment by KerfuffleV2** [comment]
> ...but, if you don't think it's worth it, by all means, feel free to close this.

I'm just a random person on the internet sharing their opinion. However, I wouldn't close it even if I could.

Regarding the roadmap though:

1. "allow parallel text generation sessions with a single model" — llama-rs already has the ability to create multiple sessions. 

2. "Extend llama_state to support loading individual model tensors." — llama-rs has its own conception of state. There's no reason it wouldn't be easy to load individual tensors.

3. "Improve quantization accuracy" — That's going to be a GGML change so llama-rs will most likely just get the benefit automatically.

4. "Try to improve threading in ggml" — Also a GGML change.

5. "Start implementing inference of other models and extend ggml operators" — The second part is a change in GGML. There are also already efforts to support other models in llama-rs. I think the main difficulty right now is that GGML doesn't support the operators required for handling the other models. Once that part is done (and especially if there's an example of using them) it shouldn't be too hard add support here.

Of course, there's certainly still room for applications (Rust or otherwise) that use llama.cpp as a library. I'd personally be less interested in them than llama-rs, because it doesn't seem like there's too much point in even writing a wrapper around a wrapper around a binding to a ML library in Rust.

**Comment by katopz** [comment]
Another random person here.
 
- Thanks @shadowmint for sharing this, I learn a lot from your POC code.
- Thanks @KerfuffleV2 for clarified, That's a lot of moving part behind the scene that I not realize there. (Wow)

For me, I love to read discussion like this, maybe we should move this to [discussion](https://github.com/rustformers/llama-rs/discussions)?

**Comment by philpax** [comment]
Closing this in favour of #126
